packages/dart/lib/src/telemetry_processing/envelope_builder.dart

```diff
@@ -0,0 +1,84 @@
+import 'package:meta/meta.dart';
+
+import '../../sentry.dart';
+import 'telemetry_buffer.dart';
+import 'telemetry_item.dart';
+
+@internal
+abstract class EnvelopeBuilder<T extends TelemetryItem> {
+  /// May return multiple envelopes (e.g., one per segment for spans).
+  List<SentryEnvelope> build(List<EncodedTelemetryItem<T>> items);
+}
+
+/// All items in a single envelope.
+class SingleEnvelopeBuilder<T extends TelemetryItem>
+    implements EnvelopeBuilder<T> {
+  final SentryEnvelope Function(List<EncodedTelemetryItem<T>> items) _create;
+
+  SingleEnvelopeBuilder(this._create);
+
+  @override
+  List<SentryEnvelope> build(List<EncodedTelemetryItem<T>> items) =>
+      items.isEmpty ? [] : [_create(items)];
+}
+
+class LogEnvelopeBuilder extends SingleEnvelopeBuilder<SentryLog> {
+  LogEnvelopeBuilder(SentryOptions options)
```

Comment:
Could we inject `sdk` instead of all options, or is it subject to change after init?